muxink/src/demux.rs

```diff
@@ -0,0 +1,254 @@
+        // Poll::Pending if its in the wrong channel, otherwise returning it in a Poll::Ready.
+        match demux.stream.poll_next_unpin(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Some(bytes)) => {
+                let channel: u8 = bytes[0..1][0];
+                let frame = bytes.slice(1..);
+                if channel == self.channel {
+                    Poll::Ready(Some(frame))
+                } else {
+                    demux.next_frame = Some((channel, frame));
+                    Poll::Pending
+                }
+            }
+            Poll::Ready(None) => Poll::Ready(None),
+        }
+
+        // TODO: figure out when we are being polled again, does it work correctly (see waker) or
+        //       will it cause inefficient races? do we need to call wake? probably. (possibly
+        //       necessary) can have table of wakers to only wake the right one.
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::marker::Unpin;
+
+    use super::*;
+    use futures::{FutureExt, Stream, StreamExt};
+
+    // This stream is used because it is not safe to call it after it returns
+    // [`Poll::Ready(None)`], whereas many other streams are. The interface for
+    // streams says that in general it is not safe, so it is important to test
+    // using a stream which has this property as well.
+    struct TestStream<T> {
+        // The items which will be returned by the stream in reverse order
+        items: Vec<T>,
+        // Once this is set to true, this `Stream` will panic upon calling [`Stream::poll_next`]
+        finished: bool,
+    }
+
+    impl<T> TestStream<T> {
+        fn new(mut items: Vec<T>) -> Self {
+            // We reverse the items as we use the pop method to remove them one by one,
+            // thus they come out in the order specified by the `Vec`.
+            items.reverse();
+            TestStream {
+                items,
+                finished: false,
+            }
+        }
+    }
+
+    // We implement Unpin because of the constraint in the implementation of the
+    // `DemultiplexerHandle`.
+    impl<T> Unpin for TestStream<T> {}
+
+    impl<T> Stream for TestStream<T> {
+        type Item = T;
+
+        fn poll_next(mut self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+            // Panic if we've already emitted [`Poll::Ready(None)`]
+            if self.finished {
+                panic!("polled a TestStream after completion");
+            }
+            if let Some(t) = self.items.pop() {
+                Poll::Ready(Some(t))
+            } else {
+                // Before we return None, make sure we set finished to true so that calling this
+                // again will result in a panic, as the specification for `Stream` tells us is
+                // possible with an arbitrary implementation.
+                self.finished = true;
+                Poll::Ready(None)
+            }
+        }
+    }
```

Comment:
We should add a simple test that ensures (via `#[should_panic]`) that this stream will indeed panic if polled after completion.